ouster_ros/src/motion_correction.cpp

```diff
@@ -0,0 +1,160 @@
+
+namespace ouster_ros {
+namespace OS1 {
+
+MotionCorrection::MotionCorrection(const size_t max_pointcloud_queue_length,
+                                   const bool output_reflectivity_as_intensity,
+                                   const bool remove_old_transformations)
+    : max_pointcloud_queue_length_(max_pointcloud_queue_length),
+      output_reflectivity_as_intensity_(output_reflectivity_as_intensity),
+      remove_old_transformations_(remove_old_transformations) {}
+
+void MotionCorrection::addTransformation(
+    const ros::Time& stamp,
+    const kindr::minimal::QuatTransformation& transformation) {
+  const uint64_t stamp_us = static_cast<uint64_t>(stamp.sec) * 1000000ul +
+                            static_cast<uint64_t>(stamp.nsec) / 1000ul;
+
+  if (!transformation_list_.empty() &&
+      stamp_us < transformation_list_.back().stamp) {
+    ROS_ERROR_STREAM(
+        "Detected jump backwards in time, resetting transformation list");
+    transformation_list_.clear();
+  }
+
+  TransformationStamped transformation_stamped;
+  transformation_stamped.transformation = transformation.cast<float>();
+  transformation_stamped.stamp = stamp_us;
+  transformation_list_.push_back(transformation_stamped);
+
+  if (transformation_list_.size() == 1) {
+    point_it_ = transformation_list_.begin();
+    cloud_it_ = transformation_list_.begin();
+  }
+}
+
+void MotionCorrection::addPointCloudToQueue(
+    const std::shared_ptr<pcl::PointCloud<PointOS1>>& pointcloud_ptr) {
+  if (!pointcloud_list_.empty() &&
+      pointcloud_ptr->header.stamp < pointcloud_list_.back()->header.stamp) {
+    ROS_ERROR_STREAM(
+        "Detected jump backwards in time, resetting pointcloud list");
+    pointcloud_list_.clear();
+  }
+
+  if (pointcloud_list_.size() > max_pointcloud_queue_length_) {
+    ROS_WARN_STREAM("There are now over "
+                    << max_pointcloud_queue_length_
+                    << " pointclouds waiting to be transformed, "
+                       "removing oldest.");
+    pointcloud_list_.pop_front();
+  }
+
+  pointcloud_list_.push_back(pointcloud_ptr);
+}
+
+bool MotionCorrection::processNextQueuedPointcloud(
+    pcl::PointCloud<pcl::PointXYZI>* pointcloud) {
+  while (!pointcloud_list_.empty()) {
+    const InterpolationStatus status =
+        transformPoints(*pointcloud_list_.front(), pointcloud);
+
+    if (status == InterpolationStatus::MATCHED) {
+      pointcloud_list_.pop_front();
+      return true;
+    }
+    if ((status == InterpolationStatus::AFTER_LAST) ||
+        (status == InterpolationStatus::NOT_ENOUGH_POINTS)) {
+      return false;
+    }
+    if (status == InterpolationStatus::BEFORE_FIRST) {
+      pointcloud_list_.pop_front();
+    }
+  }
+  return false;
+}
+
+MotionCorrection::InterpolationStatus MotionCorrection::transformPoints(
+    const pcl::PointCloud<PointOS1> pointcloud_in,
+    pcl::PointCloud<pcl::PointXYZI>* pointcloud_out) {
+  pointcloud_out->header = pointcloud_in.header;
+
```

Comment:
Here can you reserve the pointcloud points?
`pointcloud_out->reserve(pointcloud_in.size());`? Might save quite a bit of time as well.